src/vision_overhaul.cpp

```diff
@@ -0,0 +1,455 @@
+        LL_PENALTY("MinLight[LL-N]", 2);
+      } else {
+        LL_PENALTY("MinLight[LL-U]", 4);
+      }
+
+      NM_PENALTY("MinLight[NM]", 6);
+      break;
+    case LIGHT_PARTLIGHT:
+      if (has_natural_thermographic_vision) {
+        TH_PENALTY("PartLight[TH-N]", 1);
+      } else {
+        TH_PENALTY("PartLight[TH-U]", 2);
+      }
+
+      if (!has_natural_lowlight_vision) {
+        LL_PENALTY("PartLight[LL-U]", 1);
+      }
+
+      NM_PENALTY("PartLight[NM]", 2);
+      break;
+    case LIGHT_GLARE:
+      {
+        bool has_fc = FALSE;
+
+        // Find cybernetic flare comp.
+        for (struct obj_data *cyber = ch->cyberware; cyber && !has_fc; cyber = cyber->next_content) {
+          if (GET_CYBERWARE_TYPE(cyber) == CYB_EYES && IS_SET(GET_CYBERWARE_FLAGS(cyber), EYE_FLARECOMP)) {
+            has_fc = TRUE;
+            break;
+          }
+        }
+
+        // Find biologic flare comp.
+        for (struct obj_data *bio = ch->bioware; bio && !has_fc; bio = bio->next_content) {
+          if (GET_BIOWARE_TYPE(bio) == BIO_NICTATINGGLAND) {
+            has_fc = TRUE;
+            break;
+          }
+        }
+
+        // If you have a flare compensator, this doesn't affect you.
+        if (!has_fc) {
+          // You don't have a flare compensator. What a shame.
+          if (has_natural_thermographic_vision) {
+            TH_PENALTY("Glare[TH-N]", 2);
+          } else {
+            TH_PENALTY("Glare[TH-U]", 4);
+          }
+
+          if (has_natural_lowlight_vision) {
+            LL_PENALTY("Glare[LL-N]", 2);
+          } else {
+            LL_PENALTY("Glare[LL-U]", 4);
+          }
+
+          NM_PENALTY("Glare[NM]", 2);
+          break;
+        }
+      }
+      break;
+  }
+
+  if (temp_room->shadow[0]) {
+    TH_PENALTY("ShadowSpell[TH]", MIN(8, temp_room->shadow[1]));
+    LL_PENALTY("ShadowSpell[LL]", MIN(8, temp_room->shadow[1]));
+    NM_PENALTY("ShadowSpell[NM]", MIN(8, temp_room->shadow[1]));
+  }
+
+  if (light_level(temp_room) != LIGHT_GLARE && temp_room->light[1]) {
+    if (temp_room->light[2]) {
+      if (tn_with_thermo > 0) {
+        int new_thermo = MAX(0, tn_with_thermo - temp_room->light[2]);
+        if (new_thermo != tn_with_thermo) {
+          TH_PENALTY("LightSpell[TH]", new_thermo - tn_with_thermo);
+        }
+      }
+
+      if (tn_with_ll > 0) {
+        int new_ll = MAX(0, tn_with_ll - temp_room->light[2]);
+        if (new_ll != tn_with_ll) {
+          LL_PENALTY("LightSpell[LL]", new_ll - tn_with_ll);
+        }
+      }
+
+      if (tn_with_none > 0) {
+        int new_normal = MAX(0, tn_with_none - temp_room->light[2]);
+        if (new_normal != tn_with_none) {
+          NM_PENALTY("LightSpell[NM]", new_normal - tn_with_none);
+        }
+      }
+    } else {
+      // This happens when you roll no successes at all, right?
+      // light_target /= 2;
+      mudlog("Encountered that weird light case, investigate here!", ch, LOG_SYSLOG, TRUE);
```

Comment:
Expand this message as much as we can to include details to help with debugging.